### PR TITLE
[NEW] l10n_ug: Base configuration for Uganda

### PR DESCRIPTION
**Pull request by aboo-odoo**
Problem
---------
There is currently no localization for Uganda.

Objective
---------
Add the base localization for Uganda:
- Chart of accounts
- Taxes
- Fiscal positions
- Default settings
- Tax reports

Solution
---------
Create a new localization and set up all basic required information:
- tax and tax groups, CoA and fiscal positions are defined in CSVs.
Details have been obtained in documents (for details about those
documents, refer to the task).
- demo data is generated (it uses a random address)
- default accounts and value are set up for the company
- Profit and loss report and balance sheet use the generic template
- Tax report has been made following the template provided by the URA

task-3340378
---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/l10n_ug/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    "category": "Accounting/Localizations/Account Charts",
+    "license": "LGPL-3",
+    "description": """
+This is the basic Ugandian localisation necessary to run Odoo in UG:
+================================================================================
+    - Chart of accounts
+    - Taxes
+    - Fiscal positions
+    - Default settings
+    - Financial (balance sheet and profit and loss) & tax reports
```
Comment (william-andre) [inline addons/l10n_ug/__manifest__.py:16]: No financial reports in here? (and they shouldn't be)

File: addons/l10n_ug/views/report_invoice.xml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <template id="report_invoice_document" inherit_id="account.report_invoice_document">
+        <xpath expr="//div[@id='informations']" position="before">
+            <div t-if="o.company_id.country_code == 'UG'">
+                <t t-set="forced_vat" t-value="o.company_id.vat"/>
```
Comment (william-andre) [inline addons/l10n_ug/views/report_invoice.xml:6]: Is this something specific to UG?
Shouldn't the company's VAT number already be in the header/footer anyways?

Comment (aboo-odoo) [reply addons/l10n_ug/views/report_invoice.xml:6]: Indeed! I had never notice it was there. My bad!

File: addons/l10n_ug/data/account_tax_report_data.xml [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+                        </field>
+                    </record>
+                    <record id="tax_report_ug_line_F31" model="account.report.line">
+                        <field name="name">31.Input tax apportioned to Taxable Sales using selected formula</field>
+                        <field name="code">UG_TAX_F31</field>
+                        <field name="expression_ids">
+                            <record id="tax_report_ug_line_F31_tax_tag" model="account.report.expression">
+                                <field name="label">allowed</field>
+                                <field name="engine">aggregation</field>
+                                <field name="formula">(1 - UG_TAX_FM.allowed) * UG_TAX_F31.hidden_computation_of_F31_using_normal_formula + UG_TAX_FM.allowed * UG_TAX_F31.hidden_computation_of_F31_using_sam_formula</field>
```
Comment (rugo-odoo) [inline addons/l10n_ug/data/account_tax_report_data.xml:509]: We can break this line, it's a bit long :smile: (happens several time on the file)

**Review by rugo-odoo — CHANGES_REQUESTED** [review]

**Comment by rugo-odoo** [comment]
LGTM @jco-odoo

File: addons/l10n_ug/data/account_tax_report_data.xml (hunk 3 above)
Comment (william-andre) [reply addons/l10n_ug/data/account_tax_report_data.xml:509]: that's what she said

File: addons/l10n_ug/data/template/account.tax-ug.csv [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+,,,,,,,"invoice","100","tax","411722","+18% CG Sales Tax",
+,,,,,,,"refund","100","base",,"-18% CG Sales Base",
+,,,,,,,"refund","100","tax","411722","-18% CG Sales Tax",
+"purchase_capital_goods_18","18% CG","Tax on all capital goods (business assets) purchased.","18","purchase",consu,"consumption_taxes","invoice","100","base",,"+18% CG Purchases Base","True"
+,,,,,,,"invoice","100","tax","352804","+18% CG Purchases Tax",
+,,,,,,,"refund","100","base",,"-18% CG Purchases Base",
+,,,,,,,"refund","100","tax","352804","-18% CG Purchases Tax",
+"sale_deemed_capital_goods_18","18% CG Deemed","Deemed tax on all capital goods (business assets) sold.","18","sale",consu,"consumption_taxes","invoice","100","base",,"+18% Deemed CG Sales Base","False"
+,,,,,,,"invoice","100","tax","411722","+18% Deemed CG Sales Tax",
+,,,,,,,"invoice","-100","tax","411722",,
```
Comment (jco-odoo) [inline addons/l10n_ug/data/template/account.tax-ug.csv:48]: How is this tax reverse charge?

File: addons/l10n_ug/data/template/account.account-ug.csv [hunk 5]
```diff
@@ -0,0 +1,259 @@
+id,name,code,account_type,tag_ids,reconcile
+141,Property Income,141,income_other,,False
+142,Sale of goods and services,142,income,,False
+142111,Rent & rates - produced assets,142111,income,,False
```
Comment (jco-odoo) [inline addons/l10n_ug/data/template/account.account-ug.csv:4]: So sale of goods go in to 142000 and rent & rates in 142111?  (sure it is not more of a name for all accounts under 142?)  Maybe 1420 instead of 142 can avoid the confusion.

**Review by jco-odoo — COMMENTED** [review]
I still need some time for the tax report.  It seems quite complicated.

File: addons/l10n_ug/data/template/account.tax-ug.csv (hunk 4 above)
Comment (aboo-odoo) [reply addons/l10n_ug/data/template/account.tax-ug.csv:48]: Sorry, I don't understand what you mean with that question. Could you rephrase ?

File: addons/l10n_ug/data/account_tax_report_data.xml [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+                <field name="hierarchy_level">0</field>
+                <field name="children_ids">
+                    <record id="tax_report_ug_line_C1" model="account.report.line">
+                        <field name="name">1. Zero Rated Sales Local</field>
+                        <field name="code">UG_TAX_C1</field>
+                        <field name="expression_ids">
+                            <record id="tax_report_ug_line_C1_base_tag" model="account.report.expression">
+                                <field name="label">base</field>
+                                <field name="engine">tax_tags</field>
+                                <field name="formula">0% VAT Sales Base</field>
```
[the remark this answers is not in the document]
Comment (jco-odoo) [reply addons/l10n_ug/data/account_tax_report_data.xml:30]: @aboo-odoo This would mean changing it here and for all the taxes.

[the remark this answers is not in the document]
Comment (aboo-odoo) [reply addons/l10n_ug/data/account_tax_report_data.xml:30]: The tax report was made following the template by the URA : https://drive.google.com/file/d/1gzffUBVPjUXM0yft6UBaD1qMA-AVQySN/view?usp=sharing

Sorry, I don't really understand what change we would like me to perform?

[the remark this answers is not in the document]
Comment (jco-odoo) [reply addons/l10n_ug/data/account_tax_report_data.xml:30]:
```suggestion
                                <field name="formula">C1_base</field>
```

[the remark this answers is not in the document]
Comment (jco-odoo) [reply addons/l10n_ug/data/account_tax_report_data.xml:30]: to avoid having long tag names.

[the remark this answers is not in the document]
Comment (aboo-odoo) [reply addons/l10n_ug/data/account_tax_report_data.xml:30]: Oooh okay I get it, It'll be done :+1:

**Review by jco-odoo — COMMENTED** [review]
I was wondering where all those complex formulas came from and was a bit discouraged by it, but it is just how their tax report is.  See, apart from the comment of the tags and the cross_report, it seems fine.

File: addons/l10n_ug/data/account_tax_report_data.xml [hunk 7]
```diff
@@ -0,0 +1,590 @@
...
+                            </record>
+                            <record id="F31_normal_tax" model="account.report.expression">
+                                <field name="label">hidden_computation_of_F31_using_normal_formula</field>
+                                <field name="engine">aggregation</field>
+                                <field name="formula">
+                                    (UG_TAX_D20.tax - UG_TAX_F30.disallowed)
+                                    * (UG_TAX_C1.base + UG_TAX_C2.base + UG_TAX_C4.base + UG_TAX_C5.base + UG_TAX_C6.base + UG_TAX_C7.base)
+                                    / (UG_TAX_C1.base + UG_TAX_C2.base + UG_TAX_C3.base + UG_TAX_C4.base + UG_TAX_C5.base + UG_TAX_C6.base + UG_TAX_C7.base)
+                                </field>
+                                <field name="subformula">cross_report</field>
```
[the remark this answers is not in the document]
Comment (jco-odoo) [reply addons/l10n_ug/data/account_tax_report_data.xml:522]: @aboo-odoo Ok, makes sense, although we could have named it cross_section then.  (on the other hand, these sections are not that long to put on one page...  @Laura-labe )

[the remark this answers is not in the document]
Comment (aboo-odoo) [reply addons/l10n_ug/data/account_tax_report_data.xml:522]: The problem is section F which has different column names. When I asked OCO, this was the only way.

File: addons/l10n_ug/data/template/account.account-ug.csv [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+352501,Petroleum Revenue Investment Reserve,352501,asset_non_current,,False
+352599,Other Investment Fund Shares or Units,352599,asset_non_current,,False
+352701,Forwards,352701,asset_prepayments,,False
+352702,Futures,352702,asset_prepayments,,False
+352703,Options,352703,asset_prepayments,,False
+352704,Swaps,352704,asset_prepayments,,False
+3528,Account Receivable,3528,asset_receivable,,True
+352802,Staff Advances,352802,asset_receivable,,True
+352803,Other Advances,352803,asset_receivable,,True
+352804,Taxes Receivable,352804,asset_receivable,,True
```
Comment (jco-odoo) [inline addons/l10n_ug/data/template/account.account-ug.csv:213]: Needs to be current_assets.  Normally we do not have that many receivable accounts.

File: addons/l10n_ug/data/template/account.account-ug.csv [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+411315,Other Loans,411349,liability_current,,False
+4115,Investment Fund Shares or Units,4115,liability_non_current,,False
+411601,Forwards,411601,liability_current,,False
+411602,Futures,411602,liability_current,,False
+411603,Options,411603,liability_current,,False
+411604,Swaps,411604,liability_current,,False
+4117,Accounts Payable,4117,liability_payable,,True
+411711,Arrears,411711,liability_payable,,True
+411721,Trade creditors,411721,liability_payable,,True
+411722,Taxes payable,411722,liability_payable,,True
```
Comment (jco-odoo) [inline addons/l10n_ug/data/template/account.account-ug.csv:236]: This should be current liabilities and we should not have that many payable accounts.  (see below)  (and why do we have 2?)

File: addons/l10n_ug/data/template/account.account-ug.csv [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+411724,Deposits received,411724,liability_payable,,True
+411725,Advances from Other Government Units,411725,liability_payable,,True
+411726,Deferred Income,411726,liability_payable,,True
+411798,Other Payables,411798,liability_payable,,True
+4121,Currency Deposits,4121,liability_payable,,True
+412711,Arrears,412711,liability_payable,,True
+412721,Trade creditors,412721,liability_payable,,True
+412723,Taxes payable,412723,liability_payable,,True
```
Comment (jco-odoo) [inline addons/l10n_ug/data/template/account.account-ug.csv:244]: why do we have it twice?

**Review by jco-odoo — COMMENTED** [review]
@Laura-labe

**Comment by jco-odoo** [comment]
@robodoo r+ rebase-ff
